Don't store device indices separately on FutureNCCL

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* **#48501 Don't store device indices separately on FutureNCCL**
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

mrshenli reviewed (approved):
LGTM!

> (with the risk that it'll be mismatched and/or inaccurate).

When will this happen? Would I be correct if I assume users won't trigger this as `FutureNCCL` is an internal type?

lw commented:
> When will this happen?

Yes, I meant this mostly as a risk of programming error.

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]

lw committed:
Update on "Don't store device indices separately on FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL stores a set of devices (on which the tensors in the data reside) and a CUDA event for each of those devices. In fact, each event instance also already contains the device it belongs to, which means we can avoid storing that information separately (with the risk that it'll be mismatched and/or inaccurate).

Differential Revision: [D25177554](https://our.internmc.facebook.com/intern/diff/D25177554/)

[ghstack-poisoned]
